lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,377 @@
+            )
+
+    return fperm
+
+
+def header(fil, status, timestamp, hierarchy, path):
+    if status == "start":
+        fil.write(f"\n\n{status}-{timestamp}: {hierarchy} hierarchy: {path}\n")
+    else:
+        fil.write(f"\n{status}-{timestamp}: {hierarchy} hierarchy: {path}\n")
+
+
+def dump_check(fil, hierarchy_check, controller, key):
+    if controller in hierarchy_check[key]._dict:
+        fil.write(f"\t{hierarchy_check[key].get_msg()}\n")
+        output_format(fil, hierarchy_check[key]._dict[controller])
+    return 1
+
+
+def output_format(fil, controller_list):
+    for val in sorted(controller_list):
+        fil.write(f"\t\t{val}\n")
+    return 1
+
+
+class Hierarchy(object):
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+
+
+class DictOfList(object):
+    """Class used to create and add elements to Dictionary
+    """
+
+    def __init__(self, msg):
+        self._dict = OrderedDict()
+        self._msg = msg
+
+    def get_msg(self):
+        return self._msg
+
+    def add_entry(self, key, value):
+        if key in self._dict:
+            self._dict.get(key).append(value)
+        else:
+            self._dict[key] = [value]
+
+
+class ArchiveHierarchy(Hierarchy):
+    def __init__(self, name, path, config):
+        super().__init__(name, path, config)
+
+        self.controllers = list()
+        self.bad_controllers = list()
+        self.tarballs = list()
+        self.archive_check = {
+            "unexpected_dirs": DictOfList(
+                "Unexpected state directories found in this controller directory:"
+            ),
+            "subdir_status_indicators": DictOfList(""),
+            "unexpected_symlinks": DictOfList(
+                "Unexpected symlinks in controller directory:"
+            ),
+            "unexpected_objects": DictOfList(
+                "Unexpected files in controller directory:"
+            ),
+            "non_prefixes": DictOfList(
+                "Unexpected file system objects in .prefix directory:"
+            ),
+            "wrong_prefixes": DictOfList(
+                "Wrong prefix file names found in /.prefix directory:"
+            ),
+            "prefix_status_indicators": DictOfList(""),
+        }
+
+    def add_controller(self, controller):
+        self.controllers.append(controller)
+
+    def add_bad_controller(self, controller):
+        self.bad_controllers.append(controller)
+
+    def add_tarballs(self, controller):
+        self.tarballs.append(controller)
+
+    def add_unexpected_controllers(self, val, controller, dirs):
+        self.archive_check[val].add_entry(controller, dirs)
+
+    def dump(self, fil):
+        cnt = 0
+        check_h = False
+        for controller in sorted(self.controllers):
+            check_h = any(
+                [
+                    controller in self.archive_check[key]._dict
+                    or controller not in self.tarballs
```

Comment:
That's constant (it does not depend on the `key` variable of the loop in the list comprehension) so you can pull it out of the loop and avoid recalculating it.